resource.c

```diff
@@ -0,0 +1,232 @@
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * libresource is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with libresource. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * This is main file which is starting point to all interfaces provided
+ * with libresource.
+ */
+
+#include "resource.h"
+#include <malloc.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/utsname.h>
+
+#include "resource_impl.h"
+#include "resmem.h"
+#include "resnet.h"
+
+/* Allocate memory for bulk resource information and initiate it
+ * properly.
+ */
+res_blk_t *res_build_blk(int *res_ids, int res_count)
+{
+	res_unit_t *temp;
+	res_blk_t *res = NULL;
+
+	if (!res_ids || res_count <= 0) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	/* Allocate memory to hold addresses of individual resource
+	 * information.
+	 */
+	res = (res_blk_t *)
+		malloc(sizeof(res_blk_t) + (sizeof(res_unit_t *)) * res_count);
+	res->res_count = res_count;
+
+	/* Allocate and set each resource information properly */
+	for (int i = 0; i < res_count; i++) {
+		/* Allocate memory for resource info */
+		temp = (res_unit_t *)malloc(sizeof(res_unit_t));
+		memset(temp, 0, sizeof(res_unit_t));
+		temp->status = RES_STATUS_EMPTY;
+
+		/* Some resource information are big and need extra allocation.
+		 * In these cases an address is returned which hold actual
+		 * resource information.
+		 */
+		switch (res_ids[i]) {
+		case RES_MEM_ACTIVE:
+		case RES_MEM_INACTIVE:
+		case RES_MEM_AVAILABLE:
+		case RES_MEM_FREE:
+		case RES_MEM_TOTAL:
+		case RES_MEM_PAGESIZE:
+		case RES_MEM_SWAPFREE:
+		case RES_MEM_SWAPTOTAL:
+		case RES_KERN_COMPILE_TIME:
+		case RES_KERN_RELEASE:
+		case RES_NET_ALLIFSTAT:
+			break;
+
+		case RES_NET_IFSTAT:
+			temp->data.ptr = (res_net_ifstat_t *)
+				malloc(sizeof(res_net_ifstat_t));
+			break;
+
+		case RES_MEM_INFOALL:
+			temp->data.ptr = (res_mem_infoall_t *)
+				malloc(sizeof(res_mem_infoall_t));
+			break;
+
+		default:
+			eprintf("Invalid resource ID: %d", res_ids[i]);
+			while (--i >= 0)
+				free(res->res_unit[i]);
+			free(temp);
+			free(res);
+			errno = EINVAL;
+			return NULL;
+		}
+
+		temp->res_id = res_ids[i];
+		res->res_unit[i] = temp;
+	}
+
+	return res;
+}
+
+/* Free resources allocated in res_build */
+void res_destroy_blk(res_blk_t *res)
+{
+	for (int i = 0; i < res->res_count; i++) {
+		/* Some resource information are big and needed extra
+		 * Free memory allocated for that.
+		 */
+		switch (res->res_unit[i]->res_id) {
+		case RES_NET_IFSTAT:
+		case RES_MEM_INFOALL:
+			free((res->res_unit[i])->data.ptr);
```

Comment:
Should you set .ptr = NULL after free?  This is called by the caller with no locks...